src/Framework/Sdk/SdkResultItem.cs

```diff
@@ -34,13 +34,17 @@ public override bool Equals(object obj)
         {
             if (obj is SdkResultItem item &&
                    ItemSpec == item.ItemSpec &&
-                   Metadata?.Count == item.Metadata?.Count)
+                   Metadata.Count == item.Metadata?.Count)
             {
-                if (Metadata != null)
+                foreach (var kvp in Metadata)
                 {
-                    foreach (var kvp in Metadata)
+                    if (!item.Metadata.TryGetValue(kvp.Key, out var itemValue))
```

Comment:
Like @Forgind suggested, `Metadata` here can be null if the second constructor was called on `item` and null was passed for `metadata`.

We should just initialize `Metadata` to an empty set if this happens in the second constructor.

```c#
        public SdkResultItem(string itemSpec, Dictionary<string, string> metadata)
        {
            ItemSpec = itemSpec;
            Metadata = metadata ?? new();
        }
```

Reply:
Thanks! It makes sense. Currently I'm going to add unit tests first as suggested by @rainersigwald

Reply:
Why should we initialize Metadata to null? I haven't seen people complaining about it being null, and it takes up more space as not-null than null.

Reply:
I think either way we should consider making this file `#nullable enable` to formalize the decision and let the compiler complain if we get it wrong.

Reply:
So that we can make the assumption that it isn't null instead of verifying it every time we use it. I figure it's something we've commented on many times at this point, so just initialize it and be done with it.

Reply:
I'm curious if someone has found a way to rely on it being null...if so, it sounds like a bug, so I don't think it would be a big deal to break them. I'd prefer whatever is most performant.

Reply:
@rainersigwald @Forgind 

Done creating new unit test for SdkResolverItem and also done applying PR feedback, including re-adding Metadata null check.
Do we still need to have `#nullable enable` in the SdkResultItem class?

@BenVillalobos 
About this comment https://github.com/dotnet/msbuild/pull/6839#discussion_r707774092
Do we have to ensure internally that if Metadata being passed is null then we should always have a non null Metadata in the constructor?

If we are going to have this:
```
Metadata = metadata ?? new();
```
Then I'm going to change my commit to remove all of the null check for Metadata, because as far as I know previous code before my PR has no check for null then instantiate new Metadata in the constructor. 
Also CMIIW, previous code seems to always check Metadata for null value and everyone else suggesting to check for passed null value in Metadata when calling the second constructor. So I'm a bit confused, which fix I should take whether I keep the null check or having a guarantee for Metadata to be always non-null inside the second constructor.
Please confirm. 

I'm ok to add that too, if it's decided and agreed by all three of you. 🙂

Reply:
> Do we still need to have `#nullable enable` in the SdkResultItem class?

Yes; it will help ensure we don't accidentally remove the null check later if we opt for allowing Metadata to be null.

We can talk about that on Monday. I'll post a response here afterwards. Thanks for all your work on this!

Reply:
Team triage:
Let's keep the null checks (not initialize Metadata to an empty dictionary) and enable nullable. Thanks!

Reply:
@Forgind and teams,

Thanks for the heads up 🙂 
Ok, I think my next code update is just enable nullable, CMIIW.
Going to update this today.

Reply:
Might need to tweak the hashcode, but otherwise sounds right to me!